File: packages/cefari-app/src/updates.ts
import { on } from "./events.ts";
import { invokeUpdateCheck, invokeUpdateState } from "./results.ts";
import type { UpdateCheckResult, UpdateStateResult } from "./ipc.ts";
import type { Unsubscribe } from "./transport.ts";

export type UpdatesApi = {
  state(): Promise<UpdateStateResult>;
  check(): Promise<UpdateCheckResult>;
  onStateChanged(handler: (state: UpdateStateResult) => void): Unsubscribe;
};

export const updates: UpdatesApi = {
  state: (): Promise<UpdateStateResult> =>
    invokeUpdateState({ command: "updateState" }),
  check: (): Promise<UpdateCheckResult> =>
    invokeUpdateCheck({ command: "updateCheck" }),
  onStateChanged: (handler: (state: UpdateStateResult) => void): Unsubscribe =>
    on("updateStateChanged", handler),
};

File: crates/cefari-core/src/ipc/capabilities/updates.rs
capability! {
    name: updates,
    order: 40,
    support: desktopOnly,
    targets: [desktop],
    rationale: "The desktop updater contract should not be reused for app-store mobile updates.",
    commands: [
        UpdateState,
        UpdateCheck,
        UpdateApply(UpdateApplyRequest),
        UpdateRestart,
    ],
    results: [
        UpdateState(UpdateStateResult),
        UpdateCheck(UpdateCheckResult),
        UpdateApply(UpdateApplyResult),
    ],
    events: [
        UpdateStateChanged(UpdateStateResult),
    ],
}
